### ISSUE TITLE
Installing issues inside a virtual environment with pip and poetry.

### ISSUE THREAD

**Issue by t03i** [networkit/networkit#762]
I am trying to install NetworKit in a virtual environment using poetry or pip. 
However, using `poetry add networkit` or `pip install networkit`  it errors out:

```
  FAILED: CMakeFiles/networkit.dir/networkit/cpp/centrality/DynTopHarmonicCloseness.cpp.o
  /usr/bin/g++ -Dnetworkit_EXPORTS -I../../include -I../../extlibs/ttmath -I../../extlibs/tlx -O3 -DNDEBUG -fPIC  -DNETWORKIT_RELEASE_LOGGING -fopenmp -std=gnu++14 -MD -MT CMakeFiles/networkit.dir/networkit/cpp/centrality/DynTopHarmonicCloseness.cpp.o -MF CMakeFiles/networkit.dir/networkit/cpp/centrality/DynTopHarmonicCloseness.cpp.o.d -o CMakeFiles/networkit.dir/networkit/cpp/centrality/DynTopHarmonicCloseness.cpp.o -c ../../networkit/cpp/centrality/DynTopHarmonicCloseness.cpp
  In file included from ../../networkit/cpp/centrality/DynTopHarmonicCloseness.cpp:11:
  ../../include/networkit/auxiliary/PrioQueue.hpp:31:32: error: ‘numeric_limits’ is not a member of ‘std’
     31 |     const Key undefined = std::numeric_limits<Key>::max(); // TODO: make static
        |                                ^~~~~~~~~~~~~~
  ../../include/networkit/auxiliary/PrioQueue.hpp:31:50: error: expected primary-expression before ‘>’ token
     31 |     const Key undefined = std::numeric_limits<Key>::max(); // TODO: make static
        |                                                  ^
  ../../include/networkit/auxiliary/PrioQueue.hpp:31:53: error: ‘::max’ has not been declared; did you mean ‘std::max’?
     31 |     const Key undefined = std::numeric_limits<Key>::max(); // TODO: make static
        |                                                     ^~~
        |                                                     std::max
  In file included from /usr/include/c++/11.1.0/algorithm:62,
                   from ../../include/networkit/auxiliary/StringBuilder.hpp:4,
                   from ../../include/networkit/auxiliary/Log.hpp:7,
                   from ../../include/networkit/auxiliary/PrioQueue.hpp:14,
                   from ../../networkit/cpp/centrality/DynTopHarmonicCloseness.cpp:11:
  /usr/include/c++/11.1.0/bits/stl_algo.h:3467:5: note: ‘std::max’ declared here
   3467 |     max(initializer_list<_Tp> __l, _Compare __comp)
        |     ^~~
  [13/322] Building CXX object CMakeFiles/networkit.dir/networkit/cpp/coarsening/MatchingCoarsening.cpp.o
```
(python 3.9.5,  pip 21.1.2, poetry 1.1.6, gcc 11.1, clang 12.0, arch linux)

The regular install processing using `python setup.py` works but, from what I can tell, uses the global python version and global paths. Additionally, I think the linking paths are not set correctly (which might be the reason it fails). 
Using `python setup.py build_ext -i` and `python setup.py install` seems to install networkit correctly in the virtual environment (still building against the global python version). 
From what I have seen so far, CMake would support a `FindPython` which might help resolve the currently active paths? 

Do you see a way to install networkit into a virtualenv using pip/ poetry?

**Comment by avdgrinten** [comment]
I am not sure exactly why this bug shows in a virtual environment (possibly due to a different version of the C++ standard library inside/out of the environment?) but it looks like it could be fixed by including `<limits>` from PrioQueue.hpp.

**Comment by t03i** [comment]
I just double checked the standard library appears not  to be modified by the virtual environment. It only modifies PATH (extending it by the virtual env directory) and PYTHONHOME. 

I agree it looks like that would fix the issue, however, I thought this was already be the case. 
From my understanding of how pip builds modules, I think the build happens in a temporary directory and libraries build there might not be visible later on. The complete build log for `pip install networkit` in a virtualenv called `tmp` (only containing cython) can be found [here](https://gist.github.com/t03i/5987b18e1c54ec5a2da67d6c512168f8)

**Comment by michitux** [comment]
The problem here is that NetworKit is not compatible with GCC 11.1, as it has been observed [here](https://github.com/networkit/networkit/pull/753#issuecomment-849542223). I don't know why this is no problem for you outside the virtual environment, for me also a regular C++ build without the Python bindings fails.

**Comment by fabratu** [comment]
The include for `<limits>` was added recently to several files (in order to support GCC 11). If you use the current master-branch and environment with GCC 11.1, building with the mentioned commands should be no problem. 

> The regular install processing using `python setup.py` works but, from what I can tell, uses the global python version and global paths. 

Maybe you missed adding a parameter here, since calling `python setup.py` should first cythonize the Python-code and then quits (before further compiling any source code) because no additional command is given. I would be curious to know how you managed to trigger using a global Python version.

While I think this is unrelated to this issue, we currently start to use mechanism of `FindPython` for the upcoming windows build. Maybe it is a good idea to also use this mechanism in order to make detecting of environments more bullet-proof (in case that is indeed a problem).

**Comment by t03i** [comment]
Thank you for looking into it. Adding the additional includes seems to have indeed fixed the issue for the pip install. 

I tried to recreate the build using the incorrectly set global python environment but couldn't with the latest master. I suspect that this was an incorrectly set environment variable on my end. In any case I think using `FindPython` is a good way to making the build process robust. 

Further I looked into the installation using poetry. This does not yet work. My initial research yielded that Cython support through poetry is best achieved using PEP 518 and specifying build dependencies or a build script in a `pyproject.toml` file [[1]](https://github.com/python-poetry/poetry/issues/11#issuecomment-568214856) [[2]](https://www.python.org/dev/peps/pep-0518/). Since this it not really related to my initial issue, I document my progress on that part in another one.
